h2o-core/src/main/java/hex/grid/HyperSpaceWalker.java

```diff
@@ -191,9 +191,22 @@ public BaseWalker(MP params,
       _hyperParams = hyperParams;
       _paramsBuilderFactory = paramsBuilderFactory;
       _hyperParamNames = hyperParams.keySet().toArray(new String[0]);
-      _maxHyperSpaceSize = computeMaxSizeOfHyperSpace();
+      if (hyperParams.keySet().contains("constraints")) {
+        int constraintsNumber = _hyperParams.get("constraints").length;
+        Set<String> constraintHyperParams = new HashSet<String>();
+        for (int cNum = 0; cNum < constraintsNumber; cNum++) {
+          Map<String, Object[]> constraints = (Map<String, Object[]>)_hyperParams.get("constraints")[cNum];
+          for (Map.Entry<String, Object[]> p : constraints.entrySet()) {
+            constraintHyperParams.add(p.getKey());
+          }
+        }
+        _hyperParamNamesConstraint = new String[constraintHyperParams.size()];
+        constraintHyperParams.toArray(_hyperParamNamesConstraint);
+      } else 
+        _hyperParamNamesConstraint = new String[0];
```

Comment:
Also, I would suggest that you remove the groups/constraints from `_hyperParams` immediately here in the constructor. This would avoid you to have to handle this reserved key multiple times in the code later.

We can and should separate the representation at client API level which tries to be convenient for the end user (so that he can see all params at once), from the internal representation that must focus on robustness.